src/neuralnet/project.h

```diff
@@ -0,0 +1,185 @@
+//!
+typedef std::shared_ptr<std::vector<Project>> ProjectListPtr;
+
+//!
+//! \brief Read-only view of the Gridcoin project whitelist at a point in time.
+//!
+class WhitelistSnapshot
+{
+public:
+    typedef std::vector<Project>::size_type size_type;
+    typedef std::vector<Project>::iterator iterator;
+    typedef std::vector<Project>::const_iterator const_iterator;
+
+    //!
+    //! \brief Initialize a snapshot for the provided project list.
+    //!
+    //! \param projects A copy of the smart pointer to the project list.
+    //!
+    WhitelistSnapshot(const ProjectListPtr projects);
+
+    //!
+    //! \brief Returns an iterator to the beginning.
+    //!
+    const_iterator begin() const;
+
+    //!
+    //! \brief Returns an iterator to the end.
+    //!
+    const_iterator end() const;
+
+    //!
+    //! \brief Get the number of projects in the whitelist.
+    //!
+    size_type size() const;
+
+    //!
+    //! \brief Deteremine whether the whitelist contains any projects.
+    //!
+    //! This does not guarantee that the whitelist is up-to-date. The caller is
+    //! responsible for verifiying the block height.
+    //!
+    //! \return \c true if the whitelist contains at least one project.
+    //!
+    bool Populated() const;
+
+    //!
+    //! \brief Determine whether the specified project exists in the whitelist.
+    //!
+    //! \param name Project name matching the contract key.
+    //!
+    //! \return \c true if the whitelist contains a project with matching name.
+    //!
+    bool Contains(const std::string& name) const;
+
+    //!
+    //! \brief Create a snapshot copy sorted alphabetically by project name.
+    //!
+    //! \return A sorted copy of the snapshot.
+    //!
+    WhitelistSnapshot Sorted() const;
+
+private:
+    const ProjectListPtr m_projects;  //!< The set of whitelisted projects.
+};
+
+//!
+//! \brief Manages the collection of BOINC projects in the Gridcoin whitelist.
+//!
+//! An object of this class stores in memory the set of data that represents
+//! the BOINC projects in the Gridcoin whitelist. The application collects this
+//! data from network policy messages published to the network as contracts in
+//! a transaction.
+//!
+//! A \c Whitelist instance provides no direct access to the data it stores.
+//! Consumers call the \c Snapshot() method to obtain a view of the data. These
+//! \c WhitelistSnapshot instances enable read-only access to the project data
+//! as it existed at the time of their construction.
+//!
+//! This class implements copy-on-write semantics to facilitate thread-safety
+//! and to reduce overhead. It uses smart pointers to share ownership of project
+//! data with snapshots until the application mutates the data by adding or
+//! removing a project. During mutation, a \c Whitelist object copies its data
+//! and stores that behind a new smart pointer as an atomic operation, so any
+//! existing \c WhitelistSnapshot instances become obsolete.
```

Comment:
What is the use case of mutable whitelist snapshot? Only admin message processing should modify whitelist. And that happens under mutex lock.

Reply:
A snapshot is immutable. The comment about mutation applies only to the step that loads contracts from transaction messages. I'm trying to make it easier and faster to access the data by avoiding the lock on a mutex.

Reply:
If it was me, I would KISS and just copy the structure to scraper when it needs to run.

Reply:
Thanks @tomasbrod. I'm trying to KISS for the calling code that actually uses the project data. By abstracting the the details behind these classes, consumers don't need to worry about synchronization or efficient memory management. I'm working on beacon stuff that needs project data, and we could present it in other areas or features as well. The project data changes very infrequently, so I couldn't justify copying it every time something needed to use it 😄